openwisp_monitoring/device/admin.py

```diff
@@ -63,10 +63,32 @@ class CheckInline(GenericStackedInline):
     fields = ['check_type', 'is_active']
     readonly_fields = ['check_type']
 
+    def _set_check_inline(self):
+        self.fields = []
+        self.readonly_fields = ['created', 'modified']
+
     def has_add_permission(self, request, obj=None):
+        if request.user.has_perm(f'{self.model._meta.app_label}.add_check_inline'):
+            self._set_check_inline()
+            return True
+        return False
+
+    def has_view_permission(self, request, obj=None):
+        if request.user.has_perm(f'{self.model._meta.app_label}.view_check_inline'):
+            self._set_check_inline()
+            return True
+        return super(GenericStackedInline, self).has_view_permission(request, obj)
+
+    def has_change_permission(self, request, obj=None):
+        if request.user.has_perm(f'{self.model._meta.app_label}.change_check_inline'):
+            self._set_check_inline()
+            return True
         return False
 
     def has_delete_permission(self, request, obj=None):
+        if request.user.has_perm(f'{self.model._meta.app_label}.delete_check_inline'):
+            self._set_check_inline()
+            return True
         return False
```

Comment:
I would re-write this code in this way: 

```python
    fields = [
        'name',
        'is_active',
        'description',
        'check_type',
        'params',
    ]

    def get_fields(self, request, obj=None):
        if not self.has_change_permission(request, obj) or not self.has_view_permission(
            request, obj
        ):
            return ['check_type', 'is_active']
        return super().get_fields(request, obj)

    def get_readonly_fields(self, request, obj=None):
        if not self.has_change_permission(request, obj) or not self.has_view_permission(
            request, obj
        ):
            return ['check_type']
        return super().get_readonly_fields(request, obj)

    def has_add_permission(self, request, obj=None):
       # User will be able to add check from inline even if it only has permission 
       # to add a check object. 
        return super().has_add_permission(request, obj) or request.user.has_perm(
            f'{self.model._meta.app_label}.add_check_inline'
        )

    def has_change_permission(self, request, obj=None):
        return super().has_change_permission(request, obj) or request.user.has_perm(
            f'{self.model._meta.app_label}.change_check_inline'
        )

    def has_view_permission(self, request, obj=None):
        return super().has_view_permission(request, obj) or request.user.has_perm(
            f'{self.model._meta.app_label}.view_check_inline'
        )

    def has_delete_permission(self, request, obj=None):
        return super().has_delete_permission(request, obj) or request.user.has_perm(
            f'{self.model._meta.app_label}.delete_check_inline'
        )
```

I did rudimentary testing and it works. @Aryamanz29 please test it more and improve it

Reply:
thanks! Yes, it was working as expected :+1: